### PR TITLE
Ubercharge

### PR DESCRIPTION
**Pull request by ynot01**
# General Documentation

### Intent of your Pull Request

Introduce a new type of medigun

charge is displayed on the side

duration example

yes it works on simplemobs

full overlay seen here

- [x] Make sure it compiles
- [x] Make ubermedigun
- [x] Make it ubercharge
- [x] Make and apply an uber overlay so people know when themselves are invulnerable
- [x] Make and apply an uber color so people know when others are invulnerable
- [x] Make it easy to tell how much ubercharge is stored
- [x] Test ingame and make sure everything works
- [x] Put it in nuke ops with a reasonable price and balanced numbers

### Why is this change good for the game?

> This is a fun twist on the usual "click teammate and wait" type of gameplay. It'll also enable nukies to push rapidly into chokepoints that would otherwise be near death due to shotguns or whatnot.
In addition, tf2 is pretty iconic. Most people will get it off the bat, the rest will pick it up pretty fast.
-maxion12345

image

# Wiki Documentation

If while ubered:
If no one is attached to your beam it will turn off
If your beam loses connection to the other person it will turn off
If you drop it or store it will turn off
Flashing uber (switching targets) will turn it off
If it turns off it goes to 0% charge

### If there are any numerical values involved in your PR that will be relevant to a player, please note them here. 

Starts at 0% charge
When healing someone at full hp charges 1.25% per second, 80 second charge time
When healing someone hurt charges twice as fast
When ubering drains charge 12.5% per second (8 second uber)

# Changelog

:cl:  
rscadd: Adds an uber-capable medigun to the nukie uplink for 25TC
/:cl:

### CODE REVIEW

File: code/modules/projectiles/guns/misc/medbeam.dm [hunk 1]
```diff
@@ -133,3 +133,78 @@
...
+		ubercharge = 100
+
+	if(ubercharge < 0)
+		ubercharge = 0
+
+/obj/item/gun/medbeam/uber/equipped(mob/living/user)
+	..()
+	last_holder = user
+
+/obj/item/gun/medbeam/uber/proc/LoseTarget()
```
Comment (Bibby0110) [inline code/modules/projectiles/guns/misc/medbeam.dm]:
```suggestion
/obj/item/gun/medbeam/uber/LoseTarget()
```

**Review by Bibby0110 — CHANGES_REQUESTED** [review]
Typecast target to a medbeam gun in Trigger()

Autodoc

**Comment by UselessTheremin** [comment]
once this gets merged im making a tf2 medic syndi-kit and you cant stop me 🤩

**Comment by jcatjmeow** [comment]
Very funny meme image but you have not actually justified how this is fun, balanced, or worth the space in the code it occupies

**Review by jcatjmeow — CHANGES_REQUESTED** [review]
See above comment; additionally, this does not convey to someone who has never seen this before what the fuck is going on. Why can't I shoot the glowy man??????? You should add some sort of message for when you hit or shoot the guy that the projectile/weapon does no damage, to make it clear.

**Comment by ynot01** [comment]
>  You should add some sort of message for when you hit or shoot the guy that the projectile/weapon does no damage

i disagree, tf2 has done fine without any such indication, you live and you learn, the weird glowy man should be great enough indication that at least something is off

> Very funny meme image but you have not actually justified how this is fun, balanced, or worth the space in the code it occupies

fun heal gun that is balanced rationally by numbers and reasonable contingent mechanics, as you can see by reactions many people agree, if you disagree you could be constructive and suggest better numbers

**Comment by jcatjmeow** [comment]
> > You should add some sort of message for when you hit or shoot the guy that the projectile/weapon does no damage
> 
> i disagree, tf2 has done fine without any such indication, you live and you learn, the weird glowy man should be great enough indication that at least something is off

In tf2 the glowy man is a core feature of one of the 9 classes in the game. You'll see it once per side every round, plus there are a ton of voice lines and such related to it. 

Ss13 is not tf2. This is a niche item that does not see a lot of use right now and even then it's uncommon. You can't really apply tf2 logic to an entirely different game?? You honestly can't expect someone who sees this for the first or second time to understand what's going on, and given that it could be literal weeks before someone sees this item more than once, it's important to make it as unambiguous as possible.
> 
> > Very funny meme image but you have not actually justified how this is fun, balanced, or worth the space in the code it occupies
> 
> fun heal gun that is balanced rationally by numbers and reasonable contingent mechanics, as you can see by reactions many people agree, if you disagree you could be constructive and suggest better numbers

You have not actually justified how this is fun, balanced, or worth the space in the code it occupies. And no, saying that it's a "fun heal gun that is balanced" is not a justification. I'm not trying to be mean here, but you aren't selling me on this feature. This just looks like a reference for reference's sake with the balance being an afterthought. Why is this good for the game? Fill out the section properly, thank you.

**Comment by maxion12345** [comment]
jcat has never played tf2, prob sweaty overwatch main


Anyways, this is a fun twist on the usual "click teammate and wait" type of gameplay. It'll also enable nukies to push rapidly into chokepoints that would otherwise be near death due to shotguns or whatnot. War only would be a good restriction, and I am curious about holoparasite interaction. Does it also Uber the holo. Etc etc.

In addition, tf2 is pretty iconic. Most people will get it off the bat, the rest will pick it up pretty fast.

tl;dr it's gonna be fun and seems to be a great change

**Comment by ynot01** [comment]
> War only would be a good restriction

maybe, its hefty price takes up 25 of the 35TC you get from non war though and your teammates wont be able to capitalize as easy without a lot of TC/ammo (apart from the obvious osprey minigun + ubermedic combo), so i think its at a good spot

> Does it also Uber the holo. Etc etc.

only ubers the mob you are aiming at, holo interaction would probably be crossing into OP bc there's an invincibility holo

**Comment by ynot01** [comment]
> See above comment; additionally, this does not convey to someone who has never seen this before what the fuck is going on. Why can't I shoot the glowy man??????? You should add some sort of message for when you hit or shoot the guy that the projectile/weapon does no damage, to make it clear.

addressed

**Review by jcatjmeow — APPROVED** [review]

**Comment by jcatjmeow** [comment]
> jcat has never played tf2, prob sweaty overwatch main

I have around 500 hours in it, that wasn't the point. You cannot expect everyone to have played tf2, it's not reasonable to require that sort of knowledge in a completely unrelated game.

**Comment by Takahiru** [comment]
> > jcat has never played tf2, prob sweaty overwatch main
> 
> I have around 500 hours in it, that wasn't the point. You cannot expect everyone to have played tf2, it's not reasonable to require that sort of knowledge in a completely unrelated game.

To be honest, I don't think a new player will understand many of the items currently present in the game just by looking at them with face value. It will only take them one round which is after seeing this to learn what it is. We also have many items that are present in other games with similar mechanics to the original item.
